pep-0522.txt

```diff
@@ -0,0 +1,239 @@
+PEP: 522
+Title: Raise BlockingIOError in security sensitive APIs on Linux
+Version: $Revision$
+Last-Modified: $Date$
+Author: Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 16 June 2016
+Python-Version: 3.6
+
+
+Abstract
+========
+
+On Linux systems, the documentation for ``os.urandom`` currently makes the
+following contradictory promises:
+
+* to provide random numbers that are suitable for security sensitive
+  operations (such as client authentication and cryptography)
+* to provide access to the best available randomness source provided by
+  the underlying operating system
+* to present a relatively thin wrapper around the system ``/dev/urandom``
+  device
+
+This PEP proposes that in Python 3.6+ the 3rd guarantee be dropped in order to
+preserve the first two: on Linux systems that provide the ``getrandom()``
+syscall, ``os.urandom()`` would become a wrapper around that API, and raise
+``BlockingIOError`` in cases where directly accessing ``/dev/urandom/`` would
+instead return random data that may not be adequately unpredictable for use in
+security sensitive operations.
+
+As higher level abstractions over the lower level ``os.urandom()`` API, both
+``random.SystemRandom()`` and the ``secrets`` would also be documented as
+potentially raising ``BlockingIOError``.
+
+In all cases, as soon as a call to ``os.urandom()`` succeeds, all future
+calls to ``os.urandom()`` in that process will succeed (once the operating
+system random number generator is ready after system boot, it remains ready).
+
+
+Proposal
+========
+
+This PEP proposes that in Python 3.6+, ``os.urandom()`` be updated to call
+the new Linux ``getrandom()``` syscall in non-blocking mode if available and
+raise ``BlockingIOError: system random number generator is not ready`` if
+the kernel reports that the call would block.
+
+No changes are proposed for Windows or Mac OS X systems, as neither of those
+platforms provides any mechanism to run Python code before the operating
+system random number generator has been initialised.
```

Comment:
E_NOTENOUGHENTROPY? Wouldn't want the heat death of the universe to break the build ...

Sent from my iPhone

> On 16 Jun 2016, at 21:14, Donald Stufft <notifications@github.com> wrote:
> 
> In pep-0522.txt:
> 
> > +In all cases, as soon as a call to ``os.urandom()`` succeeds, all future
> > +calls to ``os.urandom()`` in that process will succeed (once the operating
> > +system random number generator is ready after system boot, it remains ready).
> > +
> > +
> > +Proposal
> > +========
> > +
> > +This PEP proposes that in Python 3.6+, ``os.urandom()`` be updated to call
> > +the new Linux ``getrandom()``` syscall in non-blocking mode if available and
> > +raise ``BlockingIOError: system random number generator is not ready`` if
> > +the kernel reports that the call would block.
> > +
> > +No changes are proposed for Windows or Mac OS X systems, as neither of those
> > +platforms provides any mechanism to run Python code before the operating
> > +system random number generator has been initialised.
> It might be reasonable to mention somewhere that OS X finds a secure sense of random so important, that it will kernel panic rather than continue to boot if it can't get enough random during it's boot process.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>